scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts

```diff
@@ -46,6 +48,15 @@ export class Goal {
       interval: opts.interval,
     });
   }
+  public static CoexistenceGoal(opts: { activityTemplate: ActivityTemplate, forEach: string }): ActivityCoexistenceGoal {
```

Comment:
Lets codegen out the activity names into their own map like ActivityTemplates so we enforce that the activity type exists.

Reply:
Completely agree, thanks for the quick review.

I've updated the type of `forEach` to be `ActivityType`, which is codegen'd to be a union of all the names of the activities:

```typescript
type ActivityType =
  | "PeelBanana"
  | "BiteBanana";

export const ActivityTypes =
  PeelBanana: "PeelBanana"
  BiteBanana: "BiteBanana";

declare global {
  var ActivityTypes: {
    PeelBanana: ActivityType,
    BiteBanana: ActivityType
  }
}
```

Does that seem like a viable solution? I think this means that the exact string will still be accepted (users are not forced to use `ActivityTypes.`). Not sure if that's a good or a bad thing.

Reply:
I think we can make an empty interface to make it a little easier, or we could even use a tagged enum;

BTW ...
```
export const ActivityTypes =
  PeelBanana: "PeelBanana"
  BiteBanana: "BiteBanana";
```
isn't valid typescript. We'd need it to be an enum or an object.

Reply:
Good catch - I wrote that part by hand instead of copying from IntelliJ, oops. Just went back and edited the snippet.

Currently mulling over what the empty interface would look like, and what we would gain from a tagged enum vs a regular enum.

Reply:
Sorry the term is String Enum. String enums have strings as values instead of numbers -> so what is actually deserialized has the proper string names instead of needing to map back from numerics to strings.

I don't think the empty interface approach makes sense (this would be similar to what we do for commands

```
interface PREHEAT_OVEN extends Command {}
```
where its merely an extension of Command that retains a type-level name vs the type aliases that `type PREHEAT_OVER` would do.

I definitely think tagged enums are the way here

Reply:
Thanks for the explanation - I've updated ActivityType to be a String enum. (I think it's kind of neat that `ActivityTypes` just becomes an alias for `ActivityType`).

I made the change in a new commit for now - I'll go squash it with the previous commit next time I rebase.